heron/packing/src/java/com/twitter/heron/packing/binpacking/FirstFitDecreasingPacking.java

```diff
@@ -279,31 +373,81 @@ public void close() {
+   * @param allocation existing packing plan
+   * @param componentsToScaleDown scale down factor for the components.
+   */
+  private void removeInstancesFromContainers(
+      ArrayList<Container> containers, Map<Integer, List<InstanceId>> allocation,
+      Map<String, Integer> componentsToScaleDown) {
+
+    ArrayList<RamRequirement> ramRequirements = getSortedRAMInstances(componentsToScaleDown);
+    for (RamRequirement ramRequirement : ramRequirements) {
+      String component = ramRequirement.getComponentName();
+      int numInstancesToRemove = -componentsToScaleDown.get(component);
+      for (int j = 0; j < numInstancesToRemove; j++) {
+        Resource instanceResource =
+            this.defaultInstanceResources.cloneWithRam(ramRequirement.getRamRequirement());
+        IdPair idPair = removeFFDInstance(containers, instanceResource, component);
+        if (idPair == null) {
+          throw new RuntimeException("Cannot perform scale down."
+              + " No more instances of component " + component + " exist"
+              + " in the containers.");
+        }
+        List<InstanceId> instances = allocation.get(idPair.getContainerId());
+        instances.remove(idPair.getInstanceId());
+        allocation.put(idPair.getContainerId(), instances);
+      }
+    }
   }
 
   /**
    * Assign a particular instance to an existing container or to a new container
    *
    * @return the container Id that incorporated the instance
    */
-  private int placeFFDInstance(ArrayList<Container> containers, Resource instanceResource) {
+  private int placeFFDInstance(ArrayList<Container> containers, Resource instanceResource,
+                               InstanceId instanceId) {
     boolean placed = false;
     int containerId = 0;
     for (int i = 0; i < containers.size() && !placed; i++) {
-      if (containers.get(i).add(instanceResource)) {
+      if (containers.get(i).add(instanceResource, instanceId)) {
         placed = true;
         containerId = i + 1;
       }
     }
     if (!placed) {
       containerId = allocateNewContainer(containers);
-      containers.get(containerId - 1).add(instanceResource);
+      containers.get(containerId - 1).add(instanceResource, instanceId);
     }
     return containerId;
   }
 
   /**
+   * Remove an instance of a particular component from the containers
+   *
+   * @return the pairId that captures the corresponding container and instance id.
+   */
+  private IdPair removeFFDInstance(ArrayList<Container> containers, Resource instanceResource,
+                                   String component) {
+    boolean removed = false;
+    int containerId = 0;
+    for (int i = 0; i < containers.size() && !removed; i++) {
+      InstanceId instanceId = containers.get(i).remove(instanceResource, component);
+      if (instanceId != null) {
+        removed = true;
+        containerId = i + 1;
+        return new IdPair(containerId, instanceId);
+      }
+    }
+    return null;
```

Comment:
Please use Optional instead of null. Or make this method return void if the caller doesn't care either way. See #1311